File: Ejercicios Semana 6/Clase de herencia/Ejercicios/Ejercicio 1.py
class Producto:
    def __init__(self, nombre, precio, categoria):
        self.nombre = nombre
        self.precio = precio
        self.categoria = categoria

    # metodo
    def mostrar_detalle(self):
        print(f"nombre: {self.nombre}")
        print(f"precio: {self.precio}")
        print(f"categoria: {self.categoria}")


class Electronico(Producto):
    def __init__(self, nombre, precio, categoria, unidades):
        super().__init__(nombre, precio, categoria)
        self.unidades = unidades

    def mostrar_detalle(self):
        print("Producto Electronico:")
        super().mostrar_detalle()
        print(f"Unidades disponibles: {self.unidades}\n")


class Alimeticio(Producto):
    def __init__(self, nombre, precio, categoria, paquetes):
        super().__init__(nombre, precio, categoria)
        self.paquetes = paquetes

    def mostrar_detalle(self):
        print("Producto Alimenticio:")
        super().mostrar_detalle()
        print(f"N de paquetes: {self.paquetes}\n")


class Vestimenta(Producto):
    def __init__(self, nombre, precio, categoria, color):
        super().__init__(nombre, precio, categoria)
        self.color = color

    def mostrar_detalle(self):
        print("Producto de vestir:")
        super().mostrar_detalle()
        print(f"Color: {self.color}")


# Instancias de cada objeto
electronico = Electronico("Lavadora", "$120.000", "Linea Blanca", 5)
print()
alimenticio = Alimeticio("Arroz", "$7.000", "Alimentos", 2)
vestimenta = Vestimenta("Pantalones", "$12.000", "Ropa", "Rojo")

productos = [electronico, alimenticio, vestimenta]
for i in productos:
    i.mostrar_detalle()

File: Ejercicios Semana 6/Clase de herencia/Ejercicios/Ejercicio 1.cpp
#include <iostream>
#include <string>

class Producto
{
public:
    std::string nombre;
    std::string precio;
    std::string categoria;

    Producto(std::string nombre, std::string precio, std::string categoria)
    {
        this->nombre = nombre;
        this->precio = precio;
        this->categoria = categoria;
    }

    void mostrar_detalle()
    {
        std::cout << "nombre: " << nombre << std::endl;
        std::cout << "precio: " << precio << std::endl;
        std::cout << "categoria: " << categoria << std::endl;
    }
};

class Electronico : public Producto
{
public:
    int unidades;

    Electronico(std::string nombre, std::string precio, std::string categoria, int unidades)
        : Producto(nombre, precio, categoria)
    {
        this->unidades = unidades;
    }

    void mostrar_detalle()
    {
        std::cout << "Producto Electronico:" << std::endl;
        Producto::mostrar_detalle();
        std::cout << "Unidades disponibles: " << unidades << std::endl;
    }
};

class Alimeticio : public Producto
{
public:
    int paquetes;

    Alimeticio(std::string nombre, std::string precio, std::string categoria, int paquetes)
        : Producto(nombre, precio, categoria)
    {
        this->paquetes = paquetes;
    }

    void mostrar_detalle()
    {
        std::cout << "Producto Alimenticio:" << std::endl;
        Producto::mostrar_detalle();
        std::cout << "N de paquetes: " << paquetes << std::endl;
    }
};

class Vestimenta : public Producto
{
public:
    std::string color;

    Vestimenta(std::string nombre, std::string precio, std::string categoria, std::string color)
        : Producto(nombre, precio, categoria)
    {
        this->color = color;
    }

    void mostrar_detalle()
    {
        std::cout << "Producto de vestir:" << std::endl;
        Producto::mostrar_detalle();
        std::cout << "Color: " << color << std::endl;
    }
};

int main()
{
    Electronico electronico("Lavadora", "$120.000", "Linea Blanca", 5);
    electronico.mostrar_detalle();
    std::cout << std::endl;

    Alimeticio alimenticio("Arroz", "$7.000", "Alimentos", 2);
    alimenticio.mostrar_detalle();
    std::cout << std::endl;

    Vestimenta vestimenta("Pantalones", "$12.000", "Ropa", "Rojo");
    vestimenta.mostrar_detalle();

    return 0;
}
